app/contracts/placeholder_users/delete_contract.rb

```diff
@@ -31,21 +31,29 @@
 module PlaceholderUsers
   class DeleteContract < ::DeleteContract
     delete_permission -> {
-      self.class.deletion_allowed?(user)
+      self.class.deletion_allowed?(model, user)
     }
 
     ##
-    # Checks if a given user may be deleted by another one.
+    # Checks if a given placeholder user may be deleted by a user.
     #
     # @param actor [User] User who wants to delete the given placeholder user.
-    def self.deletion_allowed?(actor)
-      actor.allowed_to_globally?(:manage_placeholder_user)
+    def self.deletion_allowed?(placeholder_user,
+                               actor,
+                               user_allowed_service = Authorization::UserAllowedService.new(actor))
+      actor.allowed_to_globally?(:manage_placeholder_user) &&
+        affected_projects_managed_by_actor?(placeholder_user, user_allowed_service)
```

Comment:
I'd turn this into an instance method. Doing so the `UserAllowedService` does not need to be provided but can be kept inside the DeleteContract. 

This strikes me as simpler then managing the `UserAllowedService` in the table, making it accessible to the row, with the row then passing it into the contract. 

This of course comes at the cost of instantiating the contract every time which I however find acceptable. 

The same pattern would then have to be applied to the `Users::DeleteContract`. 

In order to ease checking for deletability (and editability, writability, ...) we still need to introduce an object that can be called easily with something like:

```
can(User.current).delete?(placeholder_user)
```

or 

```
User.current.can.delete?(placeholder_user)
```

Those would in turn still call the contracts but would do so without the caller having to deal with the contracts. That change however is definitely out of scope.

Reply:
I believe I cannot follow you. The whole idea of providing the `UserAllowedService` was to reuse that service when printing rows in the placeholder users index page as its instantiation appears to be costly. Also the `UserAllowedService` instances are actor specific, which means we cannot cache it in a class variable.

Reply:
As discussed with @wielinde I was on a wrong track here since the placeholder to be deleted and by that the contract instance will change for the index action. Thus, the solution is the best we can achieve for the time being.